### PR TITLE
ARROW-5629: [C++] Fix Coverity issues

### PR DESCRIPTION
**Pull request by pitrou**
Nothing serious, just small cleanups.

Note I didn't enable all components when running the analysis, since it's quite long already. I left CUDA, Gandiva, Orc and Plasma disabled.

### CODE REVIEW

File: cpp/src/arrow/array.h [hunk 1]
```diff
@@ -529,13 +529,14 @@ class ARROW_EXPORT ListArray : public Array {
...
+  // This constructor defers SetData to a derived array class
   ListArray() = default;
   void SetData(const std::shared_ptr<ArrayData>& data);
-  const int32_t* raw_value_offsets_;
+
+  const int32_t* raw_value_offsets_ = nullptr;
 
  private:
-  const ListType* list_type_;
+  const ListType* list_type_ = nullptr;
```
Comment (fsaintjacques) [inline cpp/src/arrow/array.h]: Can't use `nullptr` in headers, use the macro `NULLPTR` instead.

Comment (pitrou) [reply cpp/src/arrow/array.h]: :-/

Comment (fsaintjacques) [reply cpp/src/arrow/array.h]: I wonder if we should keep this... https://en.wikipedia.org/wiki/C%2B%2B/CLI

**Review by bkietz — APPROVED** [review]

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/array.h]: I think that C++/CLI is the new defacto way for interfacing with C++ from C#.NET. I wrote a linter to check for nullptr and other issues, so I don't think it's a huge issue

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**